test_engine/build_test_engine.py

```diff
@@ -0,0 +1,243 @@
+    client.set_missing_host_key_policy(paramiko.AutoAddPolicy())
+    client.connect(ip, username=username, key_filename=os.path.expanduser(
+                    os.path.join("~", ".ssh", "id_rsa")), timeout=1500)
+
+    stdin, stdout, stderr = client.exec_command(commands)
+
+    while int(stdout.channel.recv_exit_status()) != 0:
+        time.sleep(RETRY_INTERVAL)
+
+    output = stdout.readlines()
+
+    client.close()
+
+    print(output)
+
+    return output
+
+
+def add_manifest(upgrade_image, version_image1, version_image2):
+
+    # upgrade image
+    commands = 'docker login -u={0} -p={1};\
+        docker manifest create {2} \\{2}-amd64 \\{2}-arm64;\
+        docker manifest push {2}'\
+            .format(docker_id, docker_pwd, upgrade_image)
+    print(commands)
+    subprocess.run(commands, shell=True)
+
+    # version image
+    commands = '\
+        docker manifest create {0} \\{0}-amd64 \\{0}-arm64;\
+        docker manifest push {0}'\
+            .format(version_image1)
+    print(commands)
+    subprocess.run(commands, shell=True)
+
+    # version image
+    commands = '\
+        docker manifest create {0} \\{0}-amd64 \\{0}-arm64;\
+        docker manifest push {0}'\
+            .format(version_image2)
+    print(commands)
+    subprocess.run(commands, shell=True)
+
+
+def generate_test_images():
+
+    # arch : amd64 arm64
+    upgrade_image, version_image1, version_image2 = get_image_names()
+    build_images("amd64", upgrade_image, version_image1, version_image2)
+    build_images("arm64", upgrade_image, version_image1, version_image2)
+    add_manifest(upgrade_image, version_image1, version_image2)
+
+
+def build_images(arch, upgrade_image, version_image1, version_image2):
+
+    os.environ['TF_VAR_build_engine_arch'] = arch
+
+    if arch == "amd64":
+
+        os.environ['TF_VAR_build_engine_aws_instance_type'] = 't2.xlarge'
+
+    elif arch == "arm64":
+
+        os.environ['TF_VAR_build_engine_aws_instance_type'] = 'a1.xlarge'
+
+    subprocess.check_output("{}/scripts/terraform_instance.sh".format(
+                            pathlib.Path(__file__).parent.absolute()))
+
+    f = open("{}/config.yml".format(pathlib.Path(__file__).parent.absolute()))
+    data = yaml.safe_load(f)
+    ip = data['nodes'][0]['address']
+    username = data['nodes'][0]['user']
+    f.close()
+
+    # upgrade test image
+    print("Start build {} upgrade test image ...".format(arch))
+    print("This may take awhile...")
+
+    commands = 'git clone https://github.com/longhorn/longhorn-engine.git;\
+            sed -i "s/.*ARG DAPPER_HOST_ARCH=.*/ARG DAPPER_HOST_ARCH={0}/" \
+                longhorn-engine/Dockerfile.dapper;\
+            cd longhorn-engine;\
+            echo -en test >> README.md;\
+            git config --global user.email "mock@gmail.com";\
+            git config --global user.name "mock";\
+            git commit -a -m "make commit number diff";\
+            sudo usermod -aG docker ubuntu;\
```

Comment:
we don't need to add ubuntu user to docker group, it has been handled in the terraform user data script for installing docker.

Reply:
I had removed this part